### ISSUE TITLE
[JENKINS-33761] Ability to disable Pipeline durability and "resume" build.

### ISSUE THREAD

**Issue by jenkins-infra-bot** [jenkinsci/workflow-cps-plugin#1172]
<p>Having some state being generated at the each node during execution, resuming builds after jenkins restarts or nodes reboots are just not feasible sometimes and can result in infinite hangs in some cases.  Also, providing durability results in extensive writes to disk that can bring performance crashing down. </p>

<p>It would be great to be able to specify that jobs don't resume upon interruptions, but rather just fail. This would increase the robustness of the system ideally, since upon nodes restarting, they quickly pick up jobs that tries to resume and hangs exhausting all available executors quickly.</p>

<p>Implementation notes:</p>
<ul>
	<li>Requires a new OptionalJobProperty on the job, optionally a new BranchProperty in workflow-multibranch-plugin that echoes that same property</li>
	<li>Needs some way to signal to storage (workflow-support) and execution (workflow-cps) that the pipeline is running with resume OFF to hint that they can use faster nondurable execution.</li>
</ul>

---
<details><summary><i>Originally reported by <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=jtilander">jtilander</a>, imported from: <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-33761" target="_blank">Ability to disable Pipeline durability and "resume" build.</a></i></summary>
<i><ul>
<li><b>assignee</b>: <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=svanoort">svanoort</a>
<li><b>status</b>: Closed
<li><b>priority</b>: Major
<li><b>component(s)</b>: workflow-api-plugin, workflow-cps-plugin, workflow-job-plugin
<li><b>label(s)</b>: project-cheetah
<li><b>resolution</b>: Fixed
<li><b>resolved</b>: 2018-01-23T01:16:02+00:00
<li><b>votes</b>: 47
<li><b>watchers</b>: 55
<li><b>imported</b>: 2025-12-07
</ul></i>
<details><summary>Raw content of original issue</summary>

<pre>
<p>Having some state being generated at the each node during execution, resuming builds after jenkins restarts or nodes reboots are just not feasible sometimes and can result in infinite hangs in some cases.  Also, providing durability results in extensive writes to disk that can bring performance crashing down. </p>

<p>It would be great to be able to specify that jobs don't resume upon interruptions, but rather just fail. This would increase the robustness of the system ideally, since upon nodes restarting, they quickly pick up jobs that tries to resume and hangs exhausting all available executors quickly.</p>

<p>Implementation notes:</p>
<ul>
	<li>Requires a new OptionalJobProperty on the job, optionally a new BranchProperty in workflow-multibranch-plugin that echoes that same property</li>
	<li>Needs some way to signal to storage (workflow-support) and execution (workflow-cps) that the pipeline is running with resume OFF to hint that they can use faster nondurable execution.</li>
</ul>
</pre>
</details>
</details>
<details><summary><i>1 attachment</i></summary>

- [JENKINS-33671_thread_dump.txt](https://raw.githubusercontent.com/jenkinsci/attachments-from-jira-issues-last/refs/heads/main/attachments/44546/JENKINS-33671_thread_dump.txt)
</details>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/jglick.png" title="jglick's avatar" /> jglick:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-33761?focusedId=255278&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-255278">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<blockquote><p>each time a job tries to resume it hangs infinitely</p></blockquote>

<p>If you can describe how to reproduce from scratch, I will try to fix it. Surviving Jenkins restarts is a key feature of Pipeline.</p>
</pre>
</details>

</ul>
</details>

<blockquote><p>each time a job tries to resume it hangs infinitely</p></blockquote>

<p>If you can describe how to reproduce from scratch, I will try to fix it. Surviving Jenkins restarts is a key feature of Pipeline.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/dodoent.png" title="dodoent's avatar" /> dodoent:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-33761?focusedId=260852&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-260852">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Well, in my case I have 7 parallel tasks running heavy shell scripts (scripts run cmake builds and ctest tests). Resuming build after restart starts the scripts from the beginning (which is basically the same as starting the whole job anew) and the worst part is that although 7 parallel branches are being executed, all executors are free, thus making possible for Jenkins to trigger another build and hog up the server resources.</p>
</pre>
</details>

</ul>
</details>

<p>Well, in my case I have 7 parallel tasks running heavy shell scripts (scripts run cmake builds and ctest tests). Resuming build after restart starts the scripts from the beginning (which is basically the same as starting the whole job anew) and the worst part is that although 7 parallel branches are being executed, all executors are free, thus making possible for Jenkins to trigger another build and hog up the server resources.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>groves:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-33761?focusedId=273640&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-273640">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>I would also really appreciate the ability to disable resumption. We have a few builds where it doesn't make sense to resume them, so it'd be better to have it off completely.</p>
</pre>
</details>

</ul>
</details>

<p>I would also really appreciate the ability to disable resumption. We have a few builds where it doesn't make sense to resume them, so it'd be better to have it off completely.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/jglick.png" title="jglick's avatar" /> jglick:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-33761?focusedId=273809&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-273809">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<blockquote><p>a few builds where it doesn't make sense to resume them</p></blockquote>

<p>Because they intrinsically could not be resumed? Or you just do not really care about loss of a build or two?</p>
</pre>
</details>

</ul>
</details>

<blockquote><p>a few builds where it doesn't make sense to resume them</p></blockquote>

<p>Because they intrinsically could not be resumed? Or you just do not really care about loss of a build or two?</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>groves:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-33761?focusedId=273865&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-273865">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>There are a couple cases:</p>
<ol>
	<li>We run builds on ephemeral EC2 agents. If Jenkins is restarted, the agents are often dead by the time Jenkins is back. Those builds just hang looking for the agent.</li>
	<li>We run deploys that require someone to monitor them. We'd prefer that they not be restarted automatically, that instead someone be there to start and watch them.</li>
</ol>

</pre>
</details>

</ul>
</details>

<p>There are a couple cases:</p>
<ol>
	<li>We run builds on ephemeral EC2 agents. If Jenkins is restarted, the agents are often dead by the time Jenkins is back. Those builds just hang looking for the agent.</li>
	<li>We run deploys that require someone to monitor them. We'd prefer that they not be restarted automatically, that instead someone be there to start and watch them.</li>
</ol>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/jglick.png" title="jglick's avatar" /> jglick:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-33761?focusedId=273870&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-273870">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<blockquote><p>We run builds on ephemeral EC2 agents. If Jenkins is restarted, the agents are often dead by the time Jenkins is back.</p></blockquote>

<p>If true, that is a bug in the EC2 plugin. It is supposed to keep the agent connected for the entire duration of the build.</p>

<blockquote><p>they not be restarted automatically</p></blockquote>

<p>Pipeline does not restart any build steps when Jenkins is restarted. It simply lets the existing process continue running and displaying output (or it might have ended on its own during the Jenkins restart).</p>
</pre>
</details>

</ul>
</details>

<blockquote><p>We run builds on ephemeral EC2 agents. If Jenkins is restarted, the agents are often dead by the time Jenkins is back.</p></blockquote>

<p>If true, that is a bug in the EC2 plugin. It is supposed to keep the agent connected for the entire duration of the build.</p>

<blockquote><p>they not be restarted automatically</p></blockquote>

<p>Pipeline does not restart any build steps when Jenkins is restarted. It simply lets the existing process continue running and displaying output (or it might have ended on its own during the Jenkins restart).</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>rg:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-33761?focusedId=273986&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-273986">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>I'd be happy with resume either working or there being an option to disable it.</p>

<p>&gt; I've also never ever seen it work correctly...</p>

<p>Likewise but maybe it happens without my noticing <img class="emoticon" src="https://issues.jenkins.io/images/icons/emoticons/smile.png" height="16" width="16" align="absmiddle" alt="" border="0"/></p>

<p>&gt; Pipeline does not restart any build steps when Jenkins is restarted. It simply lets the existing process continue running and displaying output (or it might have ended on its own during the Jenkins restart).</p>

<p>In the cases where we see resume hanging there is no process running so maybe the problem is with finding the process again, or handling it not being there.</p>

<p>Should it handle bat() and sh()? Is the assumption that the slave keeps track of the process output, return value etc.? Will a job always resume to the same slave?</p>
</pre>
</details>

</ul>
</details>

<p>I'd be happy with resume either working or there being an option to disable it.</p>

<p>> I've also never ever seen it work correctly...</p>

<p>Likewise but maybe it happens without my noticing <img class="emoticon" src="https://issues.jenkins.io/images/icons/emoticons/smile.png" height="16" width="16" align="absmiddle" alt="" border="0"/></p>

<p>> Pipeline does not restart any build steps when Jenkins is restarted. It simply lets the existing process continue running and displaying output (or it might have ended on its own during the Jenkins restart).</p>

<p>In the cases where we see resume hanging there is no process running so maybe the problem is with finding the process again, or handling it not being there.</p>

<p>Should it handle bat() and sh()? Is the assumption that the slave keeps track of the process output, return value etc.? Will a job always resume to the same slave?</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>groves:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-33761?focusedId=274037&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-274037">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<blockquote><p>If true, that is a bug in the EC2 plugin. It is supposed to keep the agent connected for the entire duration of the build.</p></blockquote>

<p>Maybe the build finished before Jenkins came back? It doesn't really change that we'd prefer not to resume these builds.</p>

<blockquote><p>Pipeline does not restart any build steps when Jenkins is restarted. It simply lets the existing process continue running and displaying output (or it might have ended on its own during the Jenkins restart).</p></blockquote>

<p>Oh, it was my understanding that it'd continue to the next step after that? Does it only complete the current step and stop?</p>
</pre>
</details>

</ul>
</details>

<blockquote><p>If true, that is a bug in the EC2 plugin. It is supposed to keep the agent connected for the entire duration of the build.</p></blockquote>

<p>Maybe the build finished before Jenkins came back? It doesn't really change that we'd prefer not to resume these builds.</p>

<blockquote><p>Pipeline does not restart any build steps when Jenkins is restarted. It simply lets the existing process continue running and displaying output (or it might have ended on its own during the Jenkins restart).</p></blockquote>

<p>Oh, it was my understanding that it'd continue to the next step after that? Does it only complete the current step and stop?</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>matthall:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-33761?focusedId=285568&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-285568">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Hello, I have recently also come across the bug of jobs not restarting, I can also provide a testcase to help with investigation, three jobs are required:</p>

<p>Job 1 will trigger job_40_sec and job_50_sec in parallel</p>

<p>If jenkins restarts or is killed when job_40_sec and job_50_sec are both running, then, when Jenkins comes back online only one of the jobs is restarted whilst the other hangs indefinitely</p>

<p>Please let me know if you need any more information or if this is the wrong place for this information</p>

<p>Pipeline scripts:</p>

<p>Job 1</p>
<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java">Map parallel_jobs = [<span class="code-quote">'branch_1'</span>: {build job: <span class="code-quote">'job_50_sec'</span>},
                     <span class="code-quote">'branch_2'</span>: {build job: <span class="code-quote">'job_40_sec'</span>}]
parallel parallel_jobs</pre>
</div></div>

<p>job_40_sec</p>
<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java">node { sleep(40) }</pre>
</div></div>

<p>job_50_sec</p>
<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java">node { sleep(50) }</pre>
</div></div>
</pre>
</details>

</ul>
</details>

<p>Hello, I have recently also come across the bug of jobs not restarting, I can also provide a testcase to help with investigation, three jobs are required:</p>

<p>Job 1 will trigger job_40_sec and job_50_sec in parallel</p>

<p>If jenkins restarts or is killed when job_40_sec and job_50_sec are both running, then, when Jenkins comes back online only one of the jobs is restarted whilst the other hangs indefinitely</p>

<p>Please let me know if you need any more information or if this is the wrong place for this information</p>

<p>Pipeline scripts:</p>

<p>Job 1</p>

<pre>
Map parallel_jobs = [<span class="code-quote">'branch_1'</span>: {build job: <span class="code-quote">'job_50_sec'</span>},
     <span class="code-quote">'branch_2'</span>: {build job: <span class="code-quote">'job_40_sec'</span>}]
parallel parallel_jobs</pre>

<p>job_40_sec</p>

<pre>
node { sleep(40) }</pre>

<p>job_50_sec</p>

<pre>
node { sleep(50) }</pre>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>rn:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-33761?focusedId=287141&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-287141">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>WRT JENKINS-41916, It would be good if the Resume build option is disabled as it doesn't respect Security.</p>
</pre>
</details>

</ul>
</details>

<p>WRT JENKINS-41916, It would be good if the Resume build option is disabled as it doesn't respect Security.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/jglick.png" title="jglick's avatar" /> jglick:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-33761?focusedId=288017&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-288017">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>FTR</p>

<blockquote><p>I have recently also come across the bug of jobs not restarting</p></blockquote>

<p>This was filed separately and a fix released.</p>
</pre>
</details>

</ul>
</details>

<p>FTR</p>

<blockquote><p>I have recently also come across the bug of jobs not restarting</p></blockquote>

<p>This was filed separately and a fix released.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>khalilj:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-33761?focusedId=302120&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-302120">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Hi,</p>

<p> </p>

<p>+1 for an option to disable pipeline resumption.</p>

<p>Thanks!</p>
</pre>
</details>

</ul>
</details>

<p>Hi,</p>

<p> </p>

<p>+1 for an option to disable pipeline resumption.</p>

<p>Thanks!</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>jeanmertz:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-33761?focusedId=304378&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-304378">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>We use the jenkins-kubernetes plugin and it also does not resume as expected.</p>

<p>Even if it did, we strive to have jobs take less than a couple of minutes, so we don't care if jobs don't resume after a restart. Having the ability to chose between "resume job", "restart job" or "discard job" would be a nice feature. We'd probably use the "restart" functionality, although I can also see use-cases for discarding the job in its entirety.</p>
</pre>
</details>

</ul>
</details>

<p>We use the jenkins-kubernetes plugin and it also does not resume as expected.</p>

<p>Even if it did, we strive to have jobs take less than a couple of minutes, so we don't care if jobs don't resume after a restart. Having the ability to chose between "resume job", "restart job" or "discard job" would be a nice feature. We'd probably use the "restart" functionality, although I can also see use-cases for discarding the job in its entirety.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/shahmishal.jpg" title="shahmishal's avatar" /> shahmishal:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-33761?focusedId=305026&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-305026">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>We should be able to disable auto resume jobs on Jenkins, it causes lots of jobs to hang for hrs before having to manually kill the builds. Also, sometime we don't get any notification about by being stuck for days because it could not resume. </p>
</pre>
</details>

</ul>
</details>

<p>We should be able to disable auto resume jobs on Jenkins, it causes lots of jobs to hang for hrs before having to manually kill the builds. Also, sometime we don't get any notification about by being stuck for days because it could not resume. </p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/maxfields2000.jpg" title="maxfields2000's avatar" /> maxfields2000:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-33761?focusedId=308989&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-308989">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Definitely need a resume disable capability. I'd love a global option, but a per job option is also not a crime. There are times when you know your workspace and what your job is doing will not be viable for resumption even if Jenkins can, in theory resume your build.  The resumes will just fail or worse lead to red-herring issues.  Yes, people should create stateless build steps with always clean thinking but that's just not reality at scale. Having a disable resume option would be handy.</p>

<p> </p>

<p>Second, if you are working in a situation where your workspace, or build nodes are ephemeral, resume literally breaks jenkins. The resume feature locks to the precise name of the comptue/build slave.  Which in a ephemeral state (think mesos/kubernetes/docker plugins, or jclouds and dynamically named slaves), when jenkins is restarted that build slave no longer exists. The job sits in the queue waiting for the slave to come online, but it never does. Because it's resume is using the slave name, not the "label" tied to the actual job, the act of being in the queue never triggers a the dynamic provisioner (jclouds) to create a new slave. And the job hangs indefinitely.</p>

<p> </p>

<p>We see evidence that these resume states can also cause thread locking on the build queue itself which then prevents any jobs from queuing at all.  We have to go through quite the arduous process to manually clean the jenkins file system to prevent builds from requeing.</p>

<p>If you won't provide s disable resume feature, t hen at least tell us the logic for how Jenkins decides which jobs need to be resumed so we can properly clean up markers jenkins looks for on the file system to tell it to requeue.</p>

<p>It seems to be a combination of jenkins home xml files as well as some file state inside the Jobs folder, (jobs/JobName/builds/#/workflow/....).  But I don't know exactly what.</p>

<p> Another option could be to wrap "node" blocks with "Resume from here" rather than resume from inside a node block.  I suppose we could try putting the "node" call inside a function marked with @NonCPS but that seems extreme and may have unexpected results and I doubt I could get all my users to follow that convention anyway.</p>

<p> </p>
</pre>
</details>

</ul>
</details>

<p>Definitely need a resume disable capability. I'd love a global option, but a per job option is also not a crime. There are times when you know your workspace and what your job is doing will not be viable for resumption even if Jenkins can, in theory resume your build.  The resumes will just fail or worse lead to red-herring issues.  Yes, people should create stateless build steps with always clean thinking but that's just not reality at scale. Having a disable resume option would be handy.</p>

<p> </p>

<p>Second, if you are working in a situation where your workspace, or build nodes are ephemeral, resume literally breaks jenkins. The resume feature locks to the precise name of the comptue/build slave.  Which in a ephemeral state (think mesos/kubernetes/docker plugins, or jclouds and dynamically named slaves), when jenkins is restarted that build slave no longer exists. The job sits in the queue waiting for the slave to come online, but it never does. Because it's resume is using the slave name, not the "label" tied to the actual job, the act of being in the queue never triggers a the dynamic provisioner (jclouds) to create a new slave. And the job hangs indefinitely.</p>

<p> </p>

<p>We see evidence that these resume states can also cause thread locking on the build queue itself which then prevents any jobs from queuing at all.  We have to go through quite the arduous process to manually clean the jenkins file system to prevent builds from requeing.</p>

<p>If you won't provide s disable resume feature, t hen at least tell us the logic for how Jenkins decides which jobs need to be resumed so we can properly clean up markers jenkins looks for on the file system to tell it to requeue.</p>

<p>It seems to be a combination of jenkins home xml files as well as some file state inside the Jobs folder, (jobs/JobName/builds/#/workflow/....).  But I don't know exactly what.</p>

<p> Another option could be to wrap "node" blocks with "Resume from here" rather than resume from inside a node block.  I suppose we could try putting the "node" call inside a function marked with @​NonCPS but that seems extreme and may have unexpected results and I doubt I could get all my users to follow that convention anyway.</p>

<p> </p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/jglick.png" title="jglick's avatar" /> jglick:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-33761?focusedId=309282&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-309282">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Please stop adding +1 comments. You may use the voting feature in JIRA.</p>
<blockquote><p>build nodes are ephemeral</p></blockquote>
<p>Already tracked as <a href="https://issue-redirect.jenkins.io/issue/36013" title="Automatically abort ExecutorPickle rehydration from an ephemeral node" class="issue-link" data-issue-key="JENKINS-36013"><del>JENKINS-36013</del></a>.</p>
</pre>
</details>

</ul>
</details>

<p>Please stop adding +1 comments. You may use the voting feature in JIRA.</p>
<blockquote><p>build nodes are ephemeral</p></blockquote>
<p>Already tracked as <a href="https://issue-redirect.jenkins.io/issue/36013" title="Automatically abort ExecutorPickle rehydration from an ephemeral node" class="issue-link" data-issue-key="JENKINS-36013"><del>JENKINS-36013</del></a>.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>svanoort:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-33761?focusedId=312853&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-312853">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=maxfields2000" class="user-hover" rel="maxfields2000">maxfields2000</a> The resume issue with missing build agents was fixed as of workflow-durable-task-step 2.15 –  patched a week or so ago.  The request for NON-durable pipeline came up at Jenkins World too, although with a different motivation (performance). </p>

<p>I think this ties into work happening now on how we store data for pipelines and their logs that I'm launching into shortly (larger project though, will take a while to land).  </p>

<p>As Jesse says, the failure to resume is generally a specific bug of some sort and needs to be addressed</p>
</pre>
</details>

</ul>
</details>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=maxfields2000" class="user-hover" rel="maxfields2000">maxfields2000</a> The resume issue with missing build agents was fixed as of workflow-durable-task-step 2.15 –  patched a week or so ago.  The request for NON-durable pipeline came up at Jenkins World too, although with a different motivation (performance). </p>

<p>I think this ties into work happening now on how we store data for pipelines and their logs that I'm launching into shortly (larger project though, will take a while to land).  </p>

<p>As Jesse says, the failure to resume is generally a specific bug of some sort and needs to be addressed</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>svanoort:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-33761?focusedId=315303&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-315303">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>I'm attaching this to the storage epic because what I have in mind will also let you use this "unsafe" mode for <b>gigantic</b> performance gains. </p>
</pre>
</details>

</ul>
</details>

<p>I'm attaching this to the storage epic because what I have in mind will also let you use this "unsafe" mode for <b>gigantic</b> performance gains. </p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/jglick.png" title="jglick's avatar" /> jglick:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-33761?focusedId=316528&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-316528">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>In such a mode you might as well also switch <tt>DurableTaskStep</tt> to be, well, nondurable, and thus to just use <tt>Launcher</tt> synchronously to run processes. Would probably require some API massaging. Would lose durability across transient agent disconnections as well as Jenkins restarts—i.e., same as traditional builds.</p>
</pre>
</details>

</ul>
</details>

<p>In such a mode you might as well also switch <tt>DurableTaskStep</tt> to be, well, nondurable, and thus to just use <tt>Launcher</tt> synchronously to run processes. Would probably require some API massaging. Would lose durability across transient agent disconnections as well as Jenkins restarts—i.e., same as traditional builds.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>svanoort:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-33761?focusedId=316648&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-316648">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=jglick" class="user-hover" rel="jglick">jglick</a> Thanks for mentioning that.  To capture an out-of-band discussion, this has value to improve reliability.  I've forked that off into <a href="https://issue-redirect.jenkins.io/issue/47390" title="Allow durable task step to, uh, not be.   Durable, that is." class="issue-link" data-issue-key="JENKINS-47390">JENKINS-47390</a>  because it is a nice work item on its own with clearly defined boundaries, and to avoid expanding the scope of this item too much. </p>
</pre>
</details>

</ul>
</details>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=jglick" class="user-hover" rel="jglick">jglick</a> Thanks for mentioning that.  To capture an out-of-band discussion, this has value to improve reliability.  I've forked that off into <a href="https://issue-redirect.jenkins.io/issue/47390" title="Allow durable task step to, uh, not be.   Durable, that is." class="issue-link" data-issue-key="JENKINS-47390">JENKINS-47390</a>  because it is a nice work item on its own with clearly defined boundaries, and to avoid expanding the scope of this item too much. </p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>sbeckwithiii:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-33761?focusedId=317414&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-317414">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>I for one would very much like the option to disable durability in pipelines. I prefer to fail early and fail fast in our environment. Pipelines with the ability to resume are fantastic in many situations but resuming in our environment continues to cause issues and headaches. At this point I am looking for ways to stop using pipelines yet have the ability to dynamically choose nodes (the only reason we use pipelines at this point is for the ability to programatically choose nodes).</p>

<p>It is far better for our automation to fail and retry again from the start or generate a report, therefore I heartily support the notion of making durability an option in pipelines rather than a requirement.</p>
</pre>
</details>

</ul>
</details>

<p>I for one would very much like the option to disable durability in pipelines. I prefer to fail early and fail fast in our environment. Pipelines with the ability to resume are fantastic in many situations but resuming in our environment continues to cause issues and headaches. At this point I am looking for ways to stop using pipelines yet have the ability to dynamically choose nodes (the only reason we use pipelines at this point is for the ability to programatically choose nodes).</p>

<p>It is far better for our automation to fail and retry again from the start or generate a report, therefore I heartily support the notion of making durability an option in pipelines rather than a requirement.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>svanoort:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-33761?focusedId=317415&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-317415">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=sbeckwithiii" class="user-hover" rel="sbeckwithiii">sbeckwithiii</a> I think you'll be happy with this when it lands - I'm hoping within the next few weeks, but take that with a grain of salt because it's part of a larger effort that brings a lot of useful features to permit pipeline to run faster and reduce the load it puts on masters.</p>

<p>I'm sorry you've had so many issues with resume though &#8211; if you wouldn't mind, could you tell us what problems you've had?  We've done a fair bit of work recently to resolve resume-time issues so it's quite likely some of these have been fixed, and if not we'd like to ensure that is robust.   Thanks!</p>
</pre>
</details>

</ul>
</details>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=sbeckwithiii" class="user-hover" rel="sbeckwithiii">sbeckwithiii</a> I think you'll be happy with this when it lands - I'm hoping within the next few weeks, but take that with a grain of salt because it's part of a larger effort that brings a lot of useful features to permit pipeline to run faster and reduce the load it puts on masters.</p>

<p>I'm sorry you've had so many issues with resume though &#8211; if you wouldn't mind, could you tell us what problems you've had?  We've done a fair bit of work recently to resolve resume-time issues so it's quite likely some of these have been fixed, and if not we'd like to ensure that is robust.   Thanks!</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>sbeckwithiii:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-33761?focusedId=317421&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-317421">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>My previous comment could be taken as a complaint or a "I expect this and deserve this" which was not my intention. We are very pleased with Jenkins as a whole. I'm surprised by how much we're able to do with it.</p>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=svanoort" class="user-hover" rel="svanoort">svanoort</a>, thank you for getting back. We have tasted some of the issues listed above including the zombie process that just would not die, as well as plugins that could use some improvement resulting in interesting indefinite wait situations like that above mentioned kubernetes plugin. My words were not meant to convey that the resumability feature of pipelines is useless or that y'all have been wasting time, but rather that I agree with others here that not every team or circumstance benefits from the resume feature and in some situations is actually hindered by it.</p>

<p>If it is useful, I can clarify why I said, "the only reason we use pipelines at this point". Why wouldn't we joyfully jump on board with pipelines and be all in? It doesn't align with the framework we've built nor does it fit the direction we've been going (wrap up automation into nice, neat little "modules" or function calls that do not require the team to learn Jenkins, Job DSL, Pipelines or Groovy), however dynamically choosing nodes based on user input is a massive boon to us in some circumstances. Where we use pipelines we must do more work to accomplish our goals and lose almost all the features of our framework but that's a choice we make. The loss is far greater to us whenever we run into the above mentioned issues with pipelines.</p>

<p>Not sure where to put this but it is worth noting that we use scripted pipelines, and do so out of necessity and design even though declarative has desirable niceties like post build steps. I didn't realize how nice it was for Jenkins to determine if a build was unstable for me until I had to write the code myself.</p>
</pre>
</details>

</ul>
</details>

<p>My previous comment could be taken as a complaint or a "I expect this and deserve this" which was not my intention. We are very pleased with Jenkins as a whole. I'm surprised by how much we're able to do with it.</p>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=svanoort" class="user-hover" rel="svanoort">svanoort</a>, thank you for getting back. We have tasted some of the issues listed above including the zombie process that just would not die, as well as plugins that could use some improvement resulting in interesting indefinite wait situations like that above mentioned kubernetes plugin. My words were not meant to convey that the resumability feature of pipelines is useless or that y'all have been wasting time, but rather that I agree with others here that not every team or circumstance benefits from the resume feature and in some situations is actually hindered by it.</p>

<p>If it is useful, I can clarify why I said, "the only reason we use pipelines at this point". Why wouldn't we joyfully jump on board with pipelines and be all in? It doesn't align with the framework we've built nor does it fit the direction we've been going (wrap up automation into nice, neat little "modules" or function calls that do not require the team to learn Jenkins, Job DSL, Pipelines or Groovy), however dynamically choosing nodes based on user input is a massive boon to us in some circumstances. Where we use pipelines we must do more work to accomplish our goals and lose almost all the features of our framework but that's a choice we make. The loss is far greater to us whenever we run into the above mentioned issues with pipelines.</p>

<p>Not sure where to put this but it is worth noting that we use scripted pipelines, and do so out of necessity and design even though declarative has desirable niceties like post build steps. I didn't realize how nice it was for Jenkins to determine if a build was unstable for me until I had to write the code myself.</p>
